### DISCUSSION TITLE
404 page after every deployment

### DISCUSSION THREAD

**Discussion by MarkwinVI** [vercel/next.js#40287] [Help] [answered]
Every time I redeploy through Vercel. If I'm already in my site having loaded the older version and then try to navigate somewhere after the deployment I get thrown a 404 page. 

The error most likely happens because of on old build file from the previous deployment, saying that they can't find the resource for the specified page. 

URL_next/data/TOeVqZ-3Rc3kVyeRo9X2J/account.json

In the example (account) is the page path, and it uses serverSideRendering. And this happens for all the pages, until I refresh the page. 

Any help would be greatly appreciated. 

I think (not sure) this started happening right after I migrated the project to next ^12.2.2

Thank you!

**Comment by icyJoseph** [comment]
I mean, this does makes sense, but somehow I feel like it shouldn't happen, or that we should be able to control it. 

Let's simplify a bit, most apps have an entry point at `page.html`, this one has a no-cache policy, and it loads scripts and links, to, among other things, download JS and CSS to run the page.

The JS and CSS is named after what you called them while developing, combined  with a hash of their contents, which means that if you don't make any changes to a JS file, and make a new build, this file in particular should be bundled and referred to, with the same name. Regardless of whether or not that holds true 100% of the time. These JS and CSS files have a caching directive, which tells the browser to store them for a long time. 

Now if I deploy version 0 of a site, and a person navigates to it, they'll get all of the JS and CSS for version 0. 

Now I deploy version 1, the person from the previous step still has HTML pointing to version 0 stuff, and more over, the JS files themselves are probably also pointing to version 0 bundles themselves. 

If the person were to reload the page they'd get an HTML file with version 1 references. However, if they stay and try to navigate or do any action, they'll be using the version 0 files, as there's no way for the client to know that upstream something has changed, without actually making a request. Moreover, it is not immediately obvious what should happen, because it depends on the deploy strategy.

If every new deploy destroys the previous artefact, then it makes sense that we start to see 404's, the files we are pointing to have disappeared. However if we just aggregate more files, then we'll get a larger output folder, but all files we need are there. Some CSR deploy strategies do this very thing.

If you have inspect the `.next` folder you'd see that it is not particularly easy to aggregate stuff.

Next.js has a mechanism called `buildId` to allow for multiple instances to run together. Think for example if your app sits behind a load balancer, and you scale by starting more next servers, then the build-id can be used to point to which ever server can handle the particular request, and keep a coherent experience for the users, https://nextjs.org/docs/api-reference/next.config.js/configuring-the-build-id.

So the question is: when a new deploy kicks in, why is the older deployment suddenly not reachable, can we control this behaviour? 

It might just be that the hobby plans do not support this, on the main deployment anyway. I'll try to read-up a bit during the day, perhaps make a deploy to see what happens.

**Reply by MarkwinVI** [reply]
Looks like this has been fixed in 12.2.5, just updated my packages, made a deployment, and now if the old build returns a 404 it performs a hard refresh on the page and loads the latest version successfully.

https://github.com/vercel/next.js/issues/39006

**Reply by icyJoseph** [reply]
There we go, so there's a mechanism upon making a request to refresh the entire page. Good to know! Good find! Could you repost your answer as a top level comment and mark that as answer? :)

**Comment by MarkwinVI** [comment, marked as answer]
Looks like this has been fixed in 12.2.5, just updated my packages, made a deployment, and now if the old build returns a 404 it performs a hard refresh on the page and loads the latest version successfully.

https://github.com/vercel/next.js/issues/39006
